File: 백준/Silver/14426. 접두사 찾기/접두사 찾기.js
const fs = require("fs");
const filePath = process.platform === "linux" ? "/dev/stdin" : "./input.txt";
let input = fs.readFileSync(filePath)
    .toString()
    .trim()
    .split("\n")
    .map(x => x.trim())

class TrieNode {
    constructor() {
        this.children = {};
        this.isLeaf = false;
    }
}

class Trie {
    constructor() {
        this.root = new TrieNode();
    }

    insert(word) {
        let node = this.root;
        for (let char of word) {
            if (!node.children[char]) {
                node.children[char] = new TrieNode();
            }
            node = node.children[char];
        }
        node.isLeaf = true;
    }

    search(word) {
        let node = this.root;
        for (let char of word) {
            if (!node.children[char]) {
                return false;
            }
            node = node.children[char];
        }
        return node.isLeaf;
    }

    startsWith(prefix) {
        let node = this.root;
        for (let char of prefix) {
            if (!node.children[char]) {
                return false;
            }
            node = node.children[char];
        }
        return true;
    }
}


function solution([nm, ...args]) {
    const [n, m] = nm.split(' ').map(Number);
    let words = args.slice(0, n);
    let prefixes = args.slice(n);
    let cnt = 0;
    let trie = new Trie();
    for(let word of words){
        trie.insert(word);
    }
    for(let prefix of prefixes){
        if(trie.startsWith(prefix)){
            cnt++;
        }
    }
    console.log(cnt);
}

solution(input);

File: 백준/Silver/14426. 접두사 찾기/접두사 찾기.kt
import java.io.BufferedReader
import java.io.File
import java.io.InputStreamReader
import java.util.*

fun main() {
    val os = System.getProperty("os.name").lowercase(Locale.getDefault())
    val args:List<String> = if (os == "linux") {
        val br = BufferedReader(InputStreamReader(System.`in`))
        br.readLines()
    } else {
        val inputFile = File("src/main/kotlin/input.txt")
        inputFile.readLines()
    }
    class TrieNode {
        val children = HashMap<Char, TrieNode>()
        var isLeaf = false
    }

    class Trie {
        private val root = TrieNode()

        fun insert(word: String) {
            var node = root
            for (char in word) {
                if (!node.children.containsKey(char)) {
                    node.children[char] = TrieNode()
                }
                node = node.children[char]!!
            }
            node.isLeaf = true
        }

        fun search(word: String): Boolean {
            var node = root
            for (char in word) {
                if (!node.children.containsKey(char)) {
                    return false
                }
                node = node.children[char]!!
            }
            return node.isLeaf
        }

        fun startsWith(prefix: String): Boolean {
            var node = root
            for (char in prefix) {
                if (!node.children.containsKey(char)) {
                    return false
                }
                node = node.children[char]!!
            }
            return true
        }
    }

    val (n, m) = args[0].split(' ').map { it.toInt() }
    val words = args.subList(1, n + 1)
    val prefixes = args.subList(n + 1, n + m + 1)
    var cnt = 0
    val trie = Trie()

    for (word in words) {
        trie.insert(word)
    }

    for (prefix in prefixes) {
        if (trie.startsWith(prefix)) {
            cnt++
        }
    }
    print(cnt)

}
